### PR TITLE
some fix first,  we would do  "optimization for limit x, y"  and "connection pool" after

### PR DESCRIPTION
**Pull request by willhan123**
1. fix cmake in windows
2. pre_sacn wolud cause failing to  mysqldump,   did not pre_scan when mysqldump
3. force_index push down, like select .. from t1 force index(k1) where xxx;
4. fix bug, share wolud init failed， cause dead loop

### CODE REVIEW

**Comment by willhan123** [comment]
for optimization for limit x,y  … 
there always has some change for mysql server.
1. add two member in THD:   longlong select_offset,  longlong select_limit;
2.  modify function: select_send::send_data, select_export::send_data, select_dump::send_data, select_dumpvar::send_data 。   Add following code in front of the function body.
  if (thd->select_offset != -1 && unit->offset_limit_cnt)
  {
    DBUG_ASSERT(thd->select_offset == 0);
    unit->offset_limit_cnt = thd->select_offset;
  }

**Comment by willhan123** [comment]
As some modifications involve changes to the mysql service layer, may you add your own  mysql server  layer  in the warehouse ?   Additional,  it will be very convenient for me to compile, debug and communicate with you.  I will look forward to it and be grateful to you.

File: CMakeLists.txt [hunk 1]
```diff
@@ -57,3 +57,10 @@ IF(ORACLE_INCLUDE_DIR AND ORACLE_OCI_LIBRARY)
     TARGET_LINK_LIBRARIES (spider ${ORACLE_OCI_LIBRARY})
   ENDIF()
 ENDIF()
+
+IF(MSVC)
+  ADD_CUSTOM_COMMAND(TARGET spider
+                     POST_BUILD 
+                     COMMAND if not exist ..\\..\\sql\\lib mkdir ..\\..\\sql\\lib\\plugin
+                     COMMAND copy Debug\\ha_spider.dll ..\\..\\sql\\lib\\plugin\\ha_spider.dll)
```
Comment (Kentoku) [inline CMakeLists.txt]: Why are this changes needed?
Is this only for "Debug" build?

File: CMakeLists.txt [hunk 2]
```diff
@@ -57,3 +57,10 @@ IF(ORACLE_INCLUDE_DIR AND ORACLE_OCI_LIBRARY)
...
+ENDIF()
```
Comment (Kentoku) [inline CMakeLists.txt]: Why are this changes needed?
Is this only for "Debug" build?

**Comment by Kentoku** [comment]
Are all changes in this pull request under the new BSD license?

File: ha_spider.cc [hunk 3]
```diff
@@ -12871,8 +12871,17 @@ int ha_spider::check_error_mode_eof(
...
   bool use_parallel
 ) {
+  THD* thd = ha_thd();
+  st_select_lex *select_lex = spider_get_select_lex(this);
   DBUG_ENTER("ha_spider::check_pre_call");
   DBUG_PRINT("info",("spider this=%p", this));
+  if (thd->lex && thd->lex->sql_command != SQLCOM_SELECT || // such like insert .. select ..
+    select_lex && select_lex->sql_cache == SELECT_LEX::SQL_NO_CACHE //  for mysqldump 
+    )
+  {
```
Comment (Kentoku) [inline ha_spider.cc:12928]: I can understand your situation, but this issue is not caused by other situation. So I'll add a parameter after merging this changes for switching this operation.
This changes are ready to merge.

File: spd_db_mysql.cc [hunk 4]
```diff
@@ -5858,6 +5858,56 @@ int spider_mysql_handler::init()
...
+        str->append(SPIDER_SQL_INDEX_USE_STR,SPIDER_SQL_INDEX_USE_LEN);
+        str->append(SPIDER_SQL_OPEN_PAREN_STR,SPIDER_SQL_OPEN_PAREN_LEN);
+        str->append(hint->key_name.str, hint->key_name.length);
+        str->append(SPIDER_SQL_CLOSE_PAREN_STR,SPIDER_SQL_CLOSE_PAREN_LEN);
+        break;
+      case INDEX_HINT_FORCE:
+        str->append(SPIDER_SQL_INDEX_FORCE_STR,SPIDER_SQL_INDEX_FORCE_LEN);
+        str->append(SPIDER_SQL_OPEN_PAREN_STR,SPIDER_SQL_OPEN_PAREN_LEN);
+        str->append(hint->key_name.str, hint->key_name.length);
+        str->append(SPIDER_SQL_CLOSE_PAREN_STR,SPIDER_SQL_CLOSE_PAREN_LEN);
```
Comment (Kentoku) [inline spd_db_mysql.cc]: Please use str->reserve() and str->q_append() for performance.

File: spd_db_mysql.cc [hunk 5]
```diff
@@ -9658,6 +9708,7 @@ int spider_mysql_handler::append_from(
     str->q_append(SPIDER_SQL_FROM_STR, SPIDER_SQL_FROM_LEN);
     table_name_pos = str->length();
     append_table_name_with_adjusting(str, link_idx, sql_type);
+    append_index_hint(str, link_idx, sql_type);
```
Comment (Kentoku) [inline spd_db_mysql.cc]: This is danger when index name is different between Spider table and data table. So I'll add a parameter after merging this changes for switching this operation. And also you can use "idx000" table parameter for adding index hints.

File: spd_db_oracle.cc [hunk 6]
```diff
@@ -5232,6 +5232,15 @@ int spider_oracle_handler::init()
   DBUG_RETURN(0);
 }
 
+int spider_oracle_handler::append_index_hint(
+  spider_string *str,
+  int link_idx,
+  ulong sql_type
+  ) 
+{
+  return 0;
```
Comment (Kentoku) [inline spd_db_oracle.cc]: Please use DBUG_ENTER() and DBUG_RETURN()

File: spd_db_oracle.cc [hunk 7]
```diff
@@ -5232,6 +5232,15 @@ int spider_oracle_handler::init()
...
+}
+
```
Comment (Kentoku) [inline spd_db_oracle.cc:5373]: Please add this function to spider_handlersocket_handler class too.

File: spd_table.cc [hunk 8]
```diff
@@ -5016,9 +5016,18 @@ SPIDER_SHARE *spider_get_share(
...
+    int sleep_cnt = 0;
     while (!share->init)
     {
-      my_sleep(10);
+      // avoid for dead loop
+      if (sleep_cnt++ > 1000)
+      {
+        fprintf(stderr, " [WARN SPIDER RESULT] "
+          "Wait share->init too long, table_name %s %s %ld\n",
+          share->table_name, share->tgt_hosts[0], share->tgt_ports[0]);
```
Comment (Kentoku) [inline spd_table.cc:5034]: Please set error_num. This is a temporary fix. I should fix more deeply after merging this changes.

File: ha_spider.cc [hunk 9]
```diff
@@ -259,6 +260,7 @@ ha_spider::ha_spider(
   result_list.tmp_tables_created = FALSE;
   result_list.bgs_working = FALSE;
   result_list.direct_order_limit = FALSE;
+  result_list.direct_limit_offset = FALSE;
```
Comment (Kentoku) [inline ha_spider.cc:269]: Please add it in ha_spider::reset() too.

File: ha_spider.cc [hunk 10]
```diff
@@ -7472,6 +7474,29 @@ int ha_spider::rnd_next_internal(
...
+      if (this->trx->thd->select_limit == 0)
+      { // mean has got all result
+        DBUG_RETURN(check_error_mode_eof(HA_ERR_END_OF_FILE));
+      }
+      longlong table_count = this->records();
+      if (table_count <= this->trx->thd->select_offset)
+      {
+        // skip this spider(partition)
+        this->trx->thd->select_offset -= table_count;
+        DBUG_RETURN(check_error_mode_eof(HA_ERR_END_OF_FILE));
```
Comment (Kentoku) [inline ha_spider.cc:7498]: Almost O.K. But I think you should better to make it without MariaDB changes.

File: CMakeLists.txt (hunk 2 above)
Comment (willhan123) [reply CMakeLists.txt]: yes,
we debug it in windows

File: ha_spider.cc (hunk 3 above)
Comment (willhan123) [reply ha_spider.cc:12928]: A parameter must work, too.
The issue happens when a query executes a long time after pre_scan.
In our scenes, only 'mysqldump' and 'insert into select' cause it.

File: spd_db_mysql.cc (hunk 4 above)
Comment (willhan123) [reply spd_db_mysql.cc]: I will fix it

File: spd_db_mysql.cc (hunk 5 above)
Comment (willhan123) [reply spd_db_mysql.cc]: I can understand and will fix it

File: spd_db_oracle.cc (hunk 6 above)
Comment (willhan123) [reply spd_db_oracle.cc]: ok

File: spd_db_oracle.cc (hunk 7 above)
Comment (willhan123) [reply spd_db_oracle.cc:5373]: ok

File: spd_table.cc (hunk 8 above)
Comment (willhan123) [reply spd_table.cc:5034]: you are more thorough

File: ha_spider.cc (hunk 9 above)
Comment (willhan123) [reply ha_spider.cc:269]: I am sorry and I will do that

File: ha_spider.cc (hunk 10 above)
Comment (willhan123) [reply ha_spider.cc:7498]: I consider whether there is a better way.
We always are confused by Mariadb's execution plan, which is not suitable in Distributed environment

File: spd_db_mysql.cc (hunk 5 above)
Comment (willhan123) [reply spd_db_mysql.cc]: I add a global parameter  'spider_index_hint_pushdown'

File: spd_conn.cc [hunk 11]
```diff
@@ -44,6 +45,9 @@ extern ulong *spd_db_att_thread_id;
 extern handlerton *spider_hton_ptr;
 extern SPIDER_DBTON spider_dbton[SPIDER_DBTON_SIZE];
 pthread_mutex_t spider_conn_id_mutex;
+pthread_mutex_t spider_ipport_count_mutex;
+pthread_mutex_t spider_conn_i_mutexs[SPIDER_MAX_PARTITION_NUM];
+pthread_cond_t  spider_conn_i_conds[SPIDER_MAX_PARTITION_NUM];
```
Comment (Kentoku) [inline spd_conn.cc:51]: Please do not use static value for it.

File: spd_conn.cc [hunk 12]
```diff
@@ -44,6 +45,9 @@ extern ulong *spd_db_att_thread_id;
 extern handlerton *spider_hton_ptr;
 extern SPIDER_DBTON spider_dbton[SPIDER_DBTON_SIZE];
 pthread_mutex_t spider_conn_id_mutex;
+pthread_mutex_t spider_ipport_count_mutex;
```
Comment (Kentoku) [inline spd_conn.cc:49]: I think this mutex can be divided by conn->key_hash_value.

File: spd_conn.cc [hunk 13]
```diff
@@ -255,6 +274,17 @@ void spider_free_conn_from_trx(
             pthread_mutex_unlock(&spider_conn_mutex);
             spider_free_conn(conn);
           } else {
+            long mutex_num=0;
+            if(ip_port_conn = (SPIDER_IP_PORT_CONN*) my_hash_search_using_hash_value(&spider_ipport_conns, conn->conn_key_hash_value, (uchar*) conn->conn_key, conn->conn_key_length))
+            {/* exists */
+              if(spider_param_max_connections())
```
Comment (Kentoku) [inline spd_conn.cc:281]: If spider_max_connections is changed to 0 dynamically, all waiting threads get timeout. I need to think about this behavior.

File: spd_conn.cc [hunk 14]
```diff
@@ -4247,3 +4328,195 @@ bool spider_conn_need_open_handler(
...
+    next_spider_conn_mutex_id = spider_conn_mutex_id;
+    if (next_spider_conn_mutex_id >= SPIDER_MAX_PARTITION_NUM) 
+    { /* RETURN ERROR and output error log*/
+      time_t cur_time = (time_t) time((time_t*) 0);
+      struct tm lt;
+      struct tm *l_time = localtime_r(&cur_time, &lt);
+      fprintf(stderr, "%04d%02d%02d %02d:%02d:%02d [WARN SPIDER RESULT] "
+                      "error to spider_create_ipport_conn, next_spider_conn_mutex_id is %d\n",
+        l_time->tm_year + 1900, l_time->tm_mon + 1, l_time->tm_mday,
+        l_time->tm_hour, l_time->tm_min, l_time->tm_sec, next_spider_conn_mutex_id);
```
Comment (Kentoku) [inline spd_conn.cc:4546]: This limitation is not good.

File: spd_include.h [hunk 15]
```diff
@@ -1325,3 +1326,17 @@ char *spider_create_string(
...
+typedef struct st_spider_ip_port_conn {
+    char *key;
+	size_t key_len;
+#ifdef SPIDER_HAS_HASH_VALUE_TYPE
+	my_hash_value_type key_hash_value;
+#endif
+    char remote_ip_str[SPIDER_CONN_META_BUF_LEN];
+    long remote_port;
+		long ip_port_count;
+		long conn_mutex_num;
```
Comment (Kentoku) [inline spd_include.h]: Please don't use TAB.

File: scripts/install_spider.sql [hunk 16]
```diff
@@ -73,7 +73,7 @@ create table if not exists mysql.spider_xa_failed_log(
 ) engine=MyISAM default charset=utf8 collate=utf8_bin;
 create table if not exists mysql.spider_tables(
   db_name char(64) not null default '',
-  table_name char(64) not null default '',
+  table_name char(256) not null default '',
```
Comment (Kentoku) [inline scripts/install_spider.sql]: Why is this 256 not 199? This change is not only for mysql.spider_tables.

File: spd_db_mysql.cc [hunk 17]
```diff
@@ -12872,6 +12886,7 @@ int spider_mysql_handler::append_order_by(
           DBUG_RETURN(HA_ERR_OUT_OF_MEM);
         str->q_append(SPIDER_SQL_COMMA_STR, SPIDER_SQL_COMMA_LEN);
       }
+*/
```
Comment (Kentoku) [inline spd_db_mysql.cc:12889]: Why is this changes needed?

File: CMakeLists.txt (hunk 2 above)
Comment (Kentoku) [reply CMakeLists.txt]: I think you can debug on windows without these changes from MariaDB 10.0. Did you try?

File: ha_spider.cc (hunk 10 above)
Comment (Kentoku) [reply ha_spider.cc:7498]: If you need a lot of time for thinking about it, I can do it after merging this changes.

**Comment by Kentoku** [comment]
@willhan123 Are all changes in this pull request under the new BSD license?

**Comment by felixliang** [comment]
Hi kentoku
All our spider patches are under the new BSD license.

Felixliang 2016.10.19

发件人: Kentoku
答复: Kentoku/Spider
日期: 2016年10月19日 星期三 02:06
至: Kentoku/Spider
抄送: Felixliang, Manual
主题: Re: [Kentoku/Spider] some fix first, we would do "optimization forlimit x, y" and "connection pool" after (#18)(Internet mail)

@willhan123https://github.com/willhan123 Are all changes in this pull request under the new BSD license?

File: CMakeLists.txt (hunk 2 above)
Comment (willhan123) [reply CMakeLists.txt]: I would try

File: spd_db_mysql.cc (hunk 17 above)
Comment (willhan123) [inline spd_db_mysql.cc]: I am sorry,

Comment (willhan123) [reply spd_db_mysql.cc:12889]: my mistake, it's not needed.
I am sorry, as I do not have your latest server layer code.  I remove it for compile.

File: spd_include.h (hunk 15 above)
Comment (willhan123) [reply spd_include.h]: I will pay attention

**Comment by Kentoku** [comment]
@felixliang Thank you for your answer.

**Comment by Kentoku** [comment]
After trying debug by Will on Windows, I'll create a branch for merging this pull request, then I'll add more changes.

File: ha_spider.cc (hunk 10 above)
Comment (willhan123) [reply ha_spider.cc:7498]: ok

File: spd_conn.cc (hunk 13 above)
Comment (willhan123) [reply spd_conn.cc:281]: after spider_get_conn_from_idle_connection return NULL (may be timeout),
if(conn==NULL && spider_max_connections ==0)
then, create conn ?
do you approve this behavior ?  I this way, some threads cost 20s seconds more.

In other way, we use a thd variables to keep spider_param_max_connections.  It must involve some changes about mysql server layer.

File: CMakeLists.txt (hunk 2 above)
Comment (willhan123) [reply CMakeLists.txt]: It  need to mkdir sql/lib/plugin in Mariadb 10.0, too
when I change some thing about spider, I also need to copy ha_spider.dll from storage/spider to sql/lib/plugin.

**Comment by felixliang** [comment]
## hi @Kentoku, will had checked the debug in MariaDB 10.0, here is his comment:

It need to mkdir sql/lib/plugin in Mariadb 10.0, too
when I change some thing about spider, I also need to copy ha_spider.dll from storage/spider to sql/lib/plugin.

**Comment by Kentoku** [comment]
I'm trying to check about this. Please wait.

**Comment by willhan123** [comment]
maria10
 yeah, i can upload the picture

**Comment by Kentoku** [comment]
Thanks.

**Comment by Kentoku** [comment]
Now this changes in develop/merge_tencent_patches_20161103 branch.

File: ha_spider.cc (hunk 3 above)
Comment (Kentoku) [reply ha_spider.cc:12928]: I just add parameters skip_parallel_search (table parameter) and spider_skip_parallel_search (server parameter). Please use "spider_skip_parallel_search = 3" for skipping pre_scan at INSERT SELECT and SQL_NO_CACHE. I wrote more detail about these parameters in commit comment.

**Comment by willhan123** [comment]
@Kentoku 
hi kentoku,
What plans do we have about 'develop/merge_tencent_patches_20161103' branch, and may I do something for you?
 
Currently, we got a spider performance problem. Do you have some good suggest ?
Query: select count(*) from tb where shard_key in ('x', 'y', 'z');
Spider would dispatch "select count(0), shard_key from tb where .... order by ...." to remote MySQL, and remote MySQL would use an inefficient index and cost lots of time.
Other way, query like 'select count(0), a from tb ' would cause 'Syntax error' in MySQL version 4.x

**Comment by Kentoku** [comment]
Hi Will,

> What plans do we have about 'develop/merge_tencent_patches_20161103' branch, and may I do something for you?

I plan to add some changes then merge it into master before Chinese New Year. If you can create some test cases (.test file), it is helpful for it.

> Currently, we got a spider performance problem. Do you have some good suggest ?
> Query: select count(*) from tb where shard_key in ('x', 'y', 'z');
> Spider would dispatch "select count(0), shard_key from tb where .... order by ...." to remote MySQL, and remote MySQL would use an inefficient index and cost lots of time.

In this case, "shard_key" is required by executor of Spider node. What kind of INDEXs are there on data node?

> Other way, query like 'select count(0), a from tb ' would cause 'Syntax error' in MySQL version 4.x

Would you please give me more detail of syntax error?

**Comment by willhan123** [comment]
@Kentoku 

Thank you!

1. I wolud create some test cases. 

2. The next picture is an example, and it take a bad index
example



3. In MySQL 4.x
select count(*), a from tb group by a.  This is  right syntax.
select count(*), a from tb.   This is a syntax error.    ERROR 1140: Mixing of GROUP columns (MIN(),MAX(),COUNT()...) with no GROUP columns is illegal if there is no GROUP BY clause

**Comment by Kentoku** [comment]
@willhan123 

> 1. I wolud create some test cases.

Thanks!

> 2. The next picture is an example, and it take a bad index example

Please get explain plan on Spider node for this query. It looks choosing bad index on Spider node.

> 3. In MySQL 4.x
> select count(), a from tb group by a. This is right syntax.
> select count(), a from tb. This is a syntax error. ERROR 1140: Mixing of GROUP columns (MIN(),MAX(),COUNT()...) with no GROUP columns is illegal if there is no GROUP BY clause

Thank you for reporting it. I'll create a strict group by mode for this case after merging.

**Comment by Kentoku** [comment]
@willhan123 @felixliang
Merging is complete. Please review my changes on develop/merge_tencent_patches_20161103 branch. If your review is OK, I'll merge this branch into master.
Sorry for delay. I'll create some tests for this patch from now.

**Comment by willhan123** [comment]
@Kentoku 
I am very glad to hear that, and  I very much agree with your changes. I wolud merge some changes to our work branch after tesing. 
By the way, when do you plan to pull request your master branch to MariaDB after merging this branch ?

**Comment by Kentoku** [comment]
@willhan123 
> By the way, when do you plan to pull request your master branch to MariaDB after merging this branch ?

It's probably a month later. It's after testing, merging to Spider master, then merging to MariaDB.

**Comment by Kentoku** [comment]
@willhan123 Now this changes were merged into Spider master.

**Comment by willhan123** [comment]
@Kentoku

**Comment by willhan123** [comment]
@Kentoku

**Comment by willhan123** [comment]
@Kentoku 
Many thanks. 
By the way, I had received the mail you sent to me,
and we will reply to you after some discussion.

**Comment by Kentoku** [comment]
@willhan123
Thanks.
Now latest Spider is testing for merging into MariaDB.
http://buildbot.askmonty.org/buildbot/grid?category=main&branch=ks-10.2-spider
